src/resolver/descriptor.cc

```diff
@@ -0,0 +1,382 @@
+              );
+            }
+          desc.path->emplace_back( * elem );
+        }
+    }
+
+  const auto & second = glob.at( 1 );
+  if ( second.has_value() &&
+       ( ( * second ) != "null" ) && ( ( * second ) != "*" )
+     )
+    {
+      desc.systems = std::vector<std::string> { * second };
+      if ( raw.systems.has_value() && ( * raw.systems != * desc.systems ) )
+        {
+          throw FloxException(
+            "`systems' list conflicts with `absPath' system specification"
+          );
+        }
+    }
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+ManifestDescriptor::ManifestDescriptor( const ManifestDescriptorRaw & raw )
+  : name( raw.name )
+  , optional( raw.optional )
+  , group( raw.packageGroup )
+{
+  /* Determine if `version' was a range or not.
+   * NOTE: The string "4.2.0" is not a range, but "4.2" is!
+   *       If you want to explicitly match the `version` field with "4.2" then
+   *       you need to use "=4.2". */
+  if ( raw.version.has_value() )
+    {
+      initManifestDescriptorVersion( * this, * raw.version );
+    }
+
+  /* You have to split `absPath' before doing most other fields. */
+  if ( raw.absPath.has_value() )
+    {
+      initManifestDescriptorAbsPath( * this, raw );
+    }
+  else if ( raw.stability.has_value() )
+    {
+      this->subtree   = Subtree( "catalog" );
+      this->stability = raw.stability;
+    }
+
+  /* Only set if it wasn't handled by `absPath`. */
+  if ( ( ! this->systems.has_value() ) && raw.systems.has_value() )
+    {
+      this->systems = * raw.systems;
+    }
+
+  if ( raw.path.has_value() )
+    {
+      /* Split relative path */
+      flox::AttrPath path;
+      if ( std::holds_alternative<std::string>( * raw.path ) )
+        {
+          path = splitAttrPath( std::get<std::string>( * raw.path ) );
+        }
+      else
+        {
+          path = std::get<AttrPath>( * raw.path );
+        }
+
+      if ( this->path.has_value() )
+        {
+          if ( this->path != path )
+            {
+              throw FloxException(
+                  "`path' conflicts with with `absPath'"
+              );
+            }
+        }
+      else
+        {
+          this->path = path;
+        }
+    }
+
+  if ( raw.packageRepository.has_value() )
+    {
+      if ( raw.input.has_value() )
+        {
+          throw FloxException(
+            "`packageRepository' may not be used with `input'"
+          );
+        }
+
+      if ( std::holds_alternative<std::string>( * raw.packageRepository ) )
+        {
+          this->input =
+            parseFlakeRef( std::get<std::string>( * raw.packageRepository ) );
+        }
+      else
+        {
+          this->input = nix::FlakeRef::fromAttrs(
+            std::get<nix::fetchers::Attrs>( * raw.packageRepository )
+          );
+        }
+      assert( std::holds_alternative<nix::FlakeRef>( * this->input ) );
```

Comment:
Could `input` just be a `FlakeRef`?

Reply:
No because it could also be a path to a nix expression file ( it can't yet, but from the perspective of the interface it is expected to in the future ).